### ISSUE TITLE
Some Build Error: Undefined Constants and Type Issues Description:

### ISSUE THREAD

**Issue by Chemple** [jmuehlig/perf-cpp#1]
Hello,

I encountered several build errors while trying to compile the project. Here are the details of the errors and my environment:

**Error Details**
When I run cmake .. followed by make -j, I receive the following errors:

```
(base) shiwen@gpu02:~/project/perf-cpp/build$ cmake ..
-- The C compiler identification is GNU 10.5.0
-- The CXX compiler identification is GNU 10.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.7s)
-- Generating done (0.0s)
-- Build files have been written to: /home/shiwen/project/perf-cpp/build
(base) shiwen@gpu02:~/project/perf-cpp/build$ make -j
[  3%] Building CXX object CMakeFiles/perf-cpp.dir/src/counter.cpp.o
[  6%] Building CXX object CMakeFiles/perf-cpp.dir/src/group.cpp.o
[ 10%] Building CXX object CMakeFiles/perf-cpp.dir/src/counter_definition.cpp.o
[ 13%] Building CXX object CMakeFiles/perf-cpp.dir/src/event_counter.cpp.o
[ 16%] Building CXX object CMakeFiles/perf-cpp.dir/src/sampler.cpp.o
/home/shiwen/project/perf-cpp/src/counter_definition.cpp: In member function ‘void perf::CounterDefinition::initialized_default_counters()’:
/home/shiwen/project/perf-cpp/src/counter_definition.cpp:62:67: error: ‘PERF_COUNT_SW_CGROUP_SWITCHES’ was not declared in this scope; did you mean ‘PERF_COUNT_SW_CONTEXT_SWITCHES’?
   62 |   this->add("cgroup-switches", CounterConfig{ PERF_TYPE_SOFTWARE, PERF_COUNT_SW_CGROUP_SWITCHES });
      |                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                   PERF_COUNT_SW_CONTEXT_SWITCHES
/home/shiwen/project/perf-cpp/src/counter_definition.cpp:62:97: error: no matching function for call to ‘perf::CounterConfig::CounterConfig(<brace-enclosed initializer list>)’
   62 |   this->add("cgroup-switches", CounterConfig{ PERF_TYPE_SOFTWARE, PERF_COUNT_SW_CGROUP_SWITCHES });
      |                                                                                                 ^
In file included from /home/shiwen/project/perf-cpp/include/perfcpp/counter_definition.h:3,
                 from /home/shiwen/project/perf-cpp/src/counter_definition.cpp:1:
/home/shiwen/project/perf-cpp/include/perfcpp/counter.h:15:3: note: candidate: ‘perf::CounterConfig::CounterConfig(uint32_t, uint64_t, uint64_t, uint64_t)’
   15 |   CounterConfig(const std::uint32_t type,
      |   ^~~~~~~~~~~~~
/home/shiwen/project/perf-cpp/include/perfcpp/counter.h:15:3: note:   conversion of argument 2 would be ill-formed:
/home/shiwen/project/perf-cpp/include/perfcpp/counter.h:12:7: note: candidate: ‘constexpr perf::CounterConfig::CounterConfig(const perf::CounterConfig&)’
   12 | class CounterConfig
      |       ^~~~~~~~~~~~~
/home/shiwen/project/perf-cpp/include/perfcpp/counter.h:12:7: note:   candidate expects 1 argument, 2 provided
In file included from /home/shiwen/project/perf-cpp/src/sampler.cpp:5:
/home/shiwen/project/perf-cpp/include/perfcpp/sampler.h:37:20: error: ‘PERF_SAMPLE_DATA_PAGE_SIZE’ was not declared in this scope; did you mean ‘PERF_SAMPLE_DATA_SRC’?
   37 |     DataPageSize = PERF_SAMPLE_DATA_PAGE_SIZE,
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                    PERF_SAMPLE_DATA_SRC
/home/shiwen/project/perf-cpp/include/perfcpp/sampler.h:38:20: error: ‘PERF_SAMPLE_CODE_PAGE_SIZE’ was not declared in this scope
   38 |     CodePageSize = PERF_SAMPLE_CODE_PAGE_SIZE,
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/shiwen/project/perf-cpp/include/perfcpp/sampler.h:39:20: error: ‘PERF_SAMPLE_WEIGHT_STRUCT’ was not declared in this scope; did you mean ‘PERF_SAMPLE_WEIGHT’?
   39 |     WeightStruct = PERF_SAMPLE_WEIGHT_STRUCT,
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                    PERF_SAMPLE_WEIGHT
/home/shiwen/project/perf-cpp/src/sampler.cpp: In member function ‘std::vector<perf::Sample> perf::Sampler::result() const’:
/home/shiwen/project/perf-cpp/src/sampler.cpp:309:40: error: ‘perf_sample_weight’ does not name a type; did you mean ‘perf_sample_regs_abi’?
  309 |         sample.weight(reinterpret_cast<perf_sample_weight*>(sample_ptr)->full);
      |                                        ^~~~~~~~~~~~~~~~~~
      |                                        perf_sample_regs_abi
/home/shiwen/project/perf-cpp/src/sampler.cpp:309:58: error: expected ‘>’ before ‘*’ token
  309 |         sample.weight(reinterpret_cast<perf_sample_weight*>(sample_ptr)->full);
      |                                                          ^
/home/shiwen/project/perf-cpp/src/sampler.cpp:309:58: error: expected ‘(’ before ‘*’ token
  309 |         sample.weight(reinterpret_cast<perf_sample_weight*>(sample_ptr)->full);
      |                                                          ^
      |                                                          (
/home/shiwen/project/perf-cpp/src/sampler.cpp:309:59: error: expected primary-expression before ‘>’ token
  309 |         sample.weight(reinterpret_cast<perf_sample_weight*>(sample_ptr)->full);
      |                                                           ^
/home/shiwen/project/perf-cpp/src/sampler.cpp:309:72: error: base operand of ‘->’ is not a pointer
  309 |         sample.weight(reinterpret_cast<perf_sample_weight*>(sample_ptr)->full);
      |                                                                        ^~
make[2]: *** [CMakeFiles/perf-cpp.dir/build.make:104: CMakeFiles/perf-cpp.dir/src/counter_definition.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/perf-cpp.dir/build.make:132: CMakeFiles/perf-cpp.dir/src/sampler.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:100: CMakeFiles/perf-cpp.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```
**Environment Details**
OS: Ubuntu 20.04
GCC: 10.5.0
CMake: 3.20.0

It appears that the errors are related to undeclared constants (PERF_COUNT_SW_CGROUP_SWITCHES, PERF_SAMPLE_DATA_PAGE_SIZE, PERF_SAMPLE_CODE_PAGE_SIZE, PERF_SAMPLE_WEIGHT_STRUCT) and type issues (perf_sample_weight). I have checked the relevant headers and it seems these constants and types are missing.

Could you please provide guidance on how to resolve these errors? If you need any further information or logs, please let me know.

Thank you for your assistance.

**Comment by jmuehlig** [comment]
Hi,
thanks for your report. Actually, your used kernel version does not provide the named constants and structs. I will implement tests that avoid those errors and allow to compile the library, even if some features are not provided. This will fix the compile-time errors, however, you will be unable to use some counters and sample features (like weight/latency sampling) with your kernel version (the library will throw exceptions if the used features are unavailable).

**Comment by jmuehlig** [comment]
Hi, I added specific tests for Kernel version `< 5.11` to the `dev` branch (commit fe18fca). Can you check if that solves the problem? In addition, could you provide the exact Kernel version of your system? Thanks!

**Comment by Chemple** [comment]
Hi,

I've updated to the latest development version and it resolved the issue successfully—thank you for the fix! By the way, the Kernel version on my system is `Linux version 5.4.0-173-generic.` Please let me know if there's anything else you need. Thanks!

**Comment by jmuehlig** [comment]
Great, thanks for the feedback. Good to know that the library works with Kernel `5.4`. I published a new version (you may want to use `0.4.1`) and updated the documentation accordingly.
